pkg/ucglib/patches/0001-add-RIOT-OS-interface.patch

```diff
@@ -0,0 +1,248 @@
+From 17bb482152c316b1bf33cd7c84a2ce96c9a720f3 Mon Sep 17 00:00:00 2001
+From: Bas Stottelaar <basstottelaar@gmail.com>
+Date: Fri, 2 Mar 2018 00:16:43 +0100
+Subject: [PATCH 1/1] add RIOT-OS interface.
+
+---
+ csrc/ucg.h        |  29 +++++++----
+ csrc/ucg_riotos.c | 152 ++++++++++++++++++++++++++++++++++++++++++++++++++++++
+ 2 files changed, 171 insertions(+), 10 deletions(-)
+ create mode 100644 csrc/ucg_riotos.c
+
+diff --git a/csrc/ucg.h b/csrc/ucg.h
+index c070a77..553562c 100644
+--- a/csrc/ucg.h
++++ b/csrc/ucg.h
+@@ -84,17 +84,19 @@
+ #include <stdint.h>
+ #include <stddef.h>
+ 
++#include "periph/gpio.h"
++
+ 
+ #ifdef __cplusplus
+ extern "C"
+ {
+ #endif
+ 
+-#if defined(ARDUINO)
+-#ifndef USE_PIN_LIST
++//#if defined(ARDUINO)
++//#ifndef USE_PIN_LIST
+ #define USE_PIN_LIST
+-#endif
+-#endif
++//#endif
++//#endif
+ 
+ #ifdef __GNUC__
+ #  define UCG_NOINLINE __attribute__((noinline))
+@@ -413,12 +415,14 @@ struct _ucg_t
+ 
+   /* only for Arduino/C++ Interface */
+ #ifdef USE_PIN_LIST
+-  uint8_t pin_list[UCG_PIN_COUNT];
++  gpio_t* pin_list;
++  uint32_t pins_enabled;
++  uint32_t dev;
+ 
+-#ifdef __AVR__
+-  volatile uint8_t *data_port[UCG_PIN_COUNT];
+-  uint8_t data_mask[UCG_PIN_COUNT];
+-#endif
++//#ifdef __AVR__
++//  volatile uint8_t *data_port[UCG_PIN_COUNT];
++//  uint8_t data_mask[UCG_PIN_COUNT];
++//#endif
+ 
+ #endif
+ 
+@@ -434,6 +438,9 @@ struct _ucg_t
+   
+ };
+ 
++#define ucg_SetPins(ucg,pins,pins_enabled) {(ucg)->pin_list = (pins); (ucg)->pins_enabled = (pins_enabled);}
++#define ucg_SetDevice(ucg,device) ((ucg)->dev = device)
++
+ #define ucg_GetWidth(ucg) ((ucg)->dimension.w)
+ #define ucg_GetHeight(ucg) ((ucg)->dimension.h)
+ 
+@@ -574,6 +581,7 @@ ucg_int_t ucg_Init(ucg_t *ucg, ucg_dev_fnptr device_cb, ucg_dev_fnptr ext_cb, uc
+ /*================================================*/
+ /* ucg_dev_sdl.c */
+ ucg_int_t ucg_sdl_dev_cb(ucg_t *ucg, ucg_int_t msg, void *data);
++int ucg_sdl_get_key(void);
+ 
+ /*================================================*/
+ /* ucg_pixel.c */
+@@ -2172,7 +2180,8 @@ extern const ucg_fntpgm_uint8_t ucg_font_osr41_tr[] UCG_FONT_SECTION("ucg_font_o
+ 
+ #endif
+ 
+-
++int16_t ucg_com_riotos_hw_spi(ucg_t *ucg, int16_t msg, uint16_t arg, uint8_t *data);
++ucg_int_t ucg_dev_dummy_cb(ucg_t *ucg, ucg_int_t msg, void *data);
+ 
+ #ifdef __cplusplus
+ }
+diff --git a/csrc/ucg_riotos.c b/csrc/ucg_riotos.c
```

Comment:
don't we put such extra files into `contrib`, or some other folder in `pkg/<name>/`, see emb6, lwip, or fatfs for instance.

Reply:
If I would put the files in `contrib` (which would be a new module), I still need to patch `ucg.h` and include the declarations over there (user should only include `ucg.h` IMHO). 

Not sure if that is a preferred option?

Reply:
I'm unsure, too - but introducing a completely new file solely for RIOT via a patch looks weird, too. At least put into a separate Patch, so that changes to the original library are self contained and the RIOT adaption (layer), too. That makes adaptions easier, I would say.

Reply:
I could also copy it before compilation, just like the Makefiles?

Reply:
@kaspar030 I think you have an opinion on this?

Reply:
> I could also copy it before compilation, just like the Makefiles?

This is what is done with [tweetnacl](https://github.com/RIOT-OS/RIOT/tree/master/pkg/tweetnacl) package, so it might be the way to go.

Reply:
I have moved the files similar to the `tweetnacl` package.